presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CrossJoinsElimination.java

```diff
@@ -98,4 +112,60 @@ public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types
         }
         return joinOrder.build().stream().map(priorities::get).collect(toImmutableList());
     }
+
+    private class Rewriter
+            extends SimplePlanRewriter<PlanNode>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final JoinGraph graph;
+        private final List<Integer> joinOrder;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator, JoinGraph graph, List<Integer> joinOrder)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.graph = requireNonNull(graph, "graph is null");
+            this.joinOrder = requireNonNull(joinOrder, "joinOrder is null");
+            checkState(joinOrder.size() >= 2);
+        }
+
+        @Override
+        public PlanNode visitPlan(PlanNode node, RewriteContext<PlanNode> context)
+        {
+            if (node.getId() != graph.getRootId()) {
+                return context.defaultRewrite(node, context.get());
+            }
+
+            PlanNode result = graph.getNode(joinOrder.get(0));
```

Comment:
I could be biased but then with such method you could write:
```
return PlanNodeSearcher.searchNode(plan)
  .where(node -> node.getId() == grapth.getRootId())
  .replaceFirst(callThatFancyMethodWhichCreatesPlanTreeWithoutCrossJoin());
```

instead of using Rewriter